File: src/engine/javascript/monteCarlo.js
import { cloneBoard, dropPiece, getAvailableColumns } from "./board";
import { simulateGame } from "./simulateGame";
const runMonteCarlo = (board, options) => {
    const { simulationsPerMove, player } = options;
    if (player === null) {
        throw new Error("Player cannot be null");
    }
    const availableColumns = getAvailableColumns(board);
    const results = [];
    for (const column of availableColumns) {
        let wins = 0;
        let losses = 0;
        let draws = 0;
        for (let simulation = 0; simulation < simulationsPerMove; simulation++) {
            const simulationBoard = cloneBoard(board);
            dropPiece(simulationBoard, column, player);
            const result = simulateGame(simulationBoard, player === "red" ? "yellow" : "red");
            if (result === "win") {
                wins++;
            }
            else if (result === "loss") {
                losses++;
            }
            else {
                draws++;
            }
        }
        const winRate = (wins / simulationsPerMove) * 100;
        results.push({
            column,
            simulations: simulationsPerMove,
            wins,
            losses,
            draws,
            winRate,
        });
        results.sort((a, b) => b.winRate - a.winRate);
    }
    return {
        bestMove: results[0]?.column ?? -1,
        results,
    };
};
export default runMonteCarlo;

File: src/engine/golang/montecarlo.go
package main

import "time"

// This file is a behavior-preserving port of
// src/engine/javascript/monteCarlo.ts.
//
// It is the public engine wrapper: AI logic stays in
// analysis.go / simulate.go / evaluate.go, and the WASM
// boundary lives separately in main.go.

// SimulationResult is the per-move Monte Carlo accounting,
// mirroring the JavaScript SimulationResult type.
type SimulationResult struct {
	Column      int
	Simulations int
	Wins        int
	Losses      int
	Draws       int
	WinRate     float64
}

// MonteCarloOptions mirrors the JavaScript MonteCarloOptions.
type MonteCarloOptions struct {
	SimulationsPerMove int
	Player             Player
}

// MonteCarloResult mirrors the JavaScript MonteCarloResult,
// plus the full analysis of the chosen move so the WASM
// boundary can return an AIEngine-compatible MoveAnalysis.
type MonteCarloResult struct {
	BestMove             int
	BestAnalysis         *MoveAnalysis
	Results              []SimulationResult
	ExecutionTime        float64 // milliseconds
	TotalSimulations     int
	SimulationsPerSecond float64
}

// RunMonteCarlo analyzes the position and returns the best
// move plus the per-column simulation results.
//
// When the best move is forcing (an immediate win, a forced
// block, or a fork), the results list is empty because the
// simulation budget was skipped.
func RunMonteCarlo(board *Board, options MonteCarloOptions) MonteCarloResult {
	start := time.Now()

	if options.Player == Empty {
		panic("Player cannot be empty")
	}

	analysis := AnalyzePosition(board, options.Player, options.SimulationsPerMove)

	bestMove := -1

	var bestAnalysis *MoveAnalysis

	if len(analysis) > 0 {
		bestMove = analysis[0].Column
		bestAnalysis = &analysis[0]
	}

	forcing := false

	if len(analysis) > 0 {
		selected := analysis[0]

		forcing = selected.IsImmediateWin ||
			selected.BlocksImmediateLoss ||
			selected.CreatesFork
	}

	executionTime := float64(time.Since(start).Microseconds()) / 1000.0

	totalSimulations := 0

	for _, result := range analysis {
		totalSimulations += result.Simulations
	}

	simulationsPerSecond := 0.0

	if totalSimulations > 0 && executionTime > 0 {
		simulationsPerSecond = float64(totalSimulations) / (executionTime / 1000)
	}

	results := []SimulationResult{}

	if !forcing {
		for _, result := range analysis {
			results = append(results, SimulationResult{
				Column:      result.Column,
				Simulations: result.Simulations,
				Wins:        result.Wins,
				Losses:      result.Losses,
				Draws:       result.Draws,
				WinRate:     result.WinRate,
			})
		}
	}

	return MonteCarloResult{
		BestMove:             bestMove,
		BestAnalysis:         bestAnalysis,
		Results:              results,
		ExecutionTime:        executionTime,
		TotalSimulations:     totalSimulations,
		SimulationsPerSecond: simulationsPerSecond,
	}
}